osu.Game/Overlays/Wiki/WikiMainPage.cs

```diff
@@ -0,0 +1,101 @@
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using HtmlAgilityPack;
+using osu.Framework.Allocation;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+
+namespace osu.Game.Overlays.Wiki
+{
+    public class WikiMainPage : FillFlowContainer
+    {
+        public string Markdown;
+
+        public WikiMainPage()
+        {
+            RelativeSizeAxes = Axes.X;
+            AutoSizeAxes = Axes.Y;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            var html = new HtmlDocument();
+            html.LoadHtml(Markdown);
+
+            var panels = createPanels(html).ToArray();
+
+            Children = new Drawable[]
+            {
+                createBlurb(html),
+                new GridContainer
+                {
+                    RelativeSizeAxes = Axes.X,
+                    AutoSizeAxes = Axes.Y,
+                    RowDimensions = Enumerable.Repeat(new Dimension(GridSizeMode.AutoSize), panels.Length).ToArray(),
+                    Content = panels,
+                },
+            };
+        }
+
+        private Container createBlurb(HtmlDocument html)
+        {
+            var blurbNode = html.DocumentNode.SelectNodes("//div[contains(@class, 'wiki-main-page__blurb')]").First();
+
+            return new Container
+            {
+                RelativeSizeAxes = Axes.X,
+                AutoSizeAxes = Axes.Y,
+                Padding = new MarginPadding
+                {
+                    Vertical = 30,
+                },
+                Child = new OsuSpriteText
+                {
+                    Text = blurbNode.InnerText,
+                    Font = OsuFont.GetFont(size: 12),
+                    Anchor = Anchor.TopCentre,
+                    Origin = Anchor.TopCentre,
+                }
+            };
+        }
+
+        private IEnumerable<Drawable[]> createPanels(HtmlDocument html)
+        {
+            var panelsNode = html.DocumentNode.SelectNodes("//div[contains(@class, 'wiki-main-page-panel')]").ToArray();
+
+            var i = 0;
+
+            while (i < panelsNode.Length)
+            {
+                var isFullWidth = panelsNode[i].HasClass("wiki-main-page-panel--full");
+
+                if (isFullWidth)
+                {
+                    yield return new Drawable[]
+                    {
+                        new WikiPanelContainer(panelsNode[i++].InnerText, true)
+                        {
+                            // This is required to fill up the space of "null" drawable below.
+                            Width = 2,
+                        },
+                        null,
+                    };
+                }
+                else
+                {
+                    yield return new Drawable[]
+                    {
+                        new WikiPanelContainer(panelsNode[i++].InnerText),
+                        new WikiPanelContainer(panelsNode[i++].InnerText),
```

Comment:
Potential index out-of-bounds here. I'll rewrite this slightly; have a better idea in mind.

Reply:
Going to leave this to you to address.

Two issues:

```
[normal] [fullwidth]
```
^ full width will not be full width

```csharp
[normal]
```
^ out of bounds on line 95

You probably want to de-nest this and construct the drawable rows on a row-by-row basis (moving the current row outside the loop).

Reply:
Okay, first, I was thinking about real world scenario about this grid. The expectation of panel grid is at least 2xn with exception of full width cells. So most likely your mentioned scenario will not likely to happen.

> You probably want to de-nest this and construct the drawable rows on a row-by-row basis (moving the current row outside the loop).

Regarding this, the data from html itself doesn't have `row` or `column` it just plain list of cell. Because in web-side itself, the styling is handled using grid css and it handle the layouting of row and column itself without need explicitly create nested row and column.

So because we only have list of cell, not list of row with the cell inside it, to do row-by-row basis, we should take 2 cells as one row manually and if it is special case of full width then the row only contains one cell. This is back to this implementation.

Reply:
In that case you can assert that that is the case in the code, and probably add some kind of failsafe (at very least a bounds check).

Reply:
I add simple bound check for now.